src/pocketmine/network/upnp/UPnP.php

```diff
@@ -19,63 +19,195 @@
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
 declare(strict_types=1);
 
 /**
- * UPnP port forwarding support. Only for Windows
+ * UPnP port forwarding support.
  */
 namespace pocketmine\network\upnp;
 
 use pocketmine\utils\Internet;
 use pocketmine\utils\Utils;
 use function class_exists;
 use function is_object;
+use const AF_INET;
+use const SOCK_DGRAM;
+use const SOL_UDP;
+use const IPPROTO_IP;
+use const IP_MULTICAST_IF;
+use const IP_MULTICAST_LOOP;
+use const IP_MULTICAST_TTL;
+use const SOL_SOCKET;
+use const SO_RCVTIMEO;
 
 abstract class UPnP{
+	/** @var string */
+	private static $serviceURL;
+
+	/**
+	 * Search the device node with given type.
+	 *
+	 * @param \SimpleXMLElement $device
+	 * @param string $deviceType
+	 * @param \SimpleXMLElement $foundDevice reference parameter
+	 *
+	 * @return bool
+	 */
+	private static function GetChildDevice(\SimpleXMLElement $device, string $deviceType, &$foundDevice) : bool{
+		foreach($device->deviceList->device as $child){
+			if((string)$child->deviceType === $deviceType){
+				$foundDevice = $child;
+				return true;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Search the service node with given type.
+	 *
+	 * @param \SimpleXMLElement $device
+	 * @param string $serviceType
+	 * @param \SimpleXMLElement $foundService reference parameter
+	 *
+	 * @return bool
+	 */
+	private static function GetChildService(\SimpleXMLElement $device, string $serviceType, &$foundService) : bool{
+		foreach($device->serviceList->service as $child){
+			if((string)$child->serviceType === $serviceType){
+				$foundService = $child;
+				return true;
+			}
+		}
+		return false;
+	}
+	
+	public static function Setup() : void{
+		$socket = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP);
+		socket_set_option($socket, SOL_SOCKET, SO_RCVTIMEO, array("sec"=>3,"usec"=>0));
+		$contents =
+			"M-SEARCH * HTTP/1.1\r\n".
+			"MX: 2\r\n".
+			"HOST: 239.255.255.250:1900\r\n".
+			"MAN: \"ssdp:discover\"\r\n".
+			"ST: upnp:rootdevice\r\n\r\n";
+		socket_sendto($socket, $contents, strlen($contents), 0, "239.255.255.250", 1900);
+		socket_recvfrom($socket, $buffer, 1024, 0, $responseHost, $responsePort);
+
+		if(!$buffer || !preg_match('/location\s*:\s*(.+)\n/i', $buffer, $matches)){
+			throw new \RuntimeException("Unable to find the router. Ensure that network discovery is enabled in Control Panel.");
+		}
+		$location = trim($matches[1]);
+		$url = parse_url($location);
+		$response = Internet::getURL($location, 3, [], $err, $headers, $httpCode);
+		if(!$response){
+			throw new \RuntimeException("Unable to access XML: {$err}");
+		}
+		if($httpCode !== 200){
+			throw new \RuntimeException("Unable to access XML: {$response}");
+		}
+		$root = new \SimpleXMLElement($response);
```

Comment:
A dependency on `ext-simplexml` needs to be added to `composer.json` for this.

Reply:
This is going to spew a bunch of warnings and an `Exception` if the data isn't valid, which might crash the server. We want a nice error to be logged, we don't want a crash.